src/Components/WebView/WebView/src/WebViewManager.cs

```diff
@@ -112,6 +112,23 @@ public Task AddRootComponentAsync(Type componentType, string selector, Parameter
         }
     }
 
+    /// <summary>
+    /// Calls the specified <paramref name="action"/> and passes in the scoped services available to Blazor components.
+    /// </summary>
+    /// <param name="action">The action to invoke.</param>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="action"/> is <c>null</c>.</exception>
+    /// <exception cref="InvalidOperationException">Thrown if this method is called before Blazor has started in the web view.</exception>
+    public void Dispatch(Action<IServiceProvider> action)
+    {
+        ArgumentNullException.ThrowIfNull(action);
+        if (_currentPageContext is null)
+        {
+            throw new InvalidOperationException($"{nameof(Dispatch)} can't be called until after Blazor has started in the web view.");
+        }
+
+        action(_currentPageContext.ServiceProvider);
+    }
```

Comment:
This solves the issue of accessing the service provider, but there's still also the issue that the `PageContext` itself could change while the action is running. For example, the user might reload the page inside the webview, and then any `IJSRuntime` you get from this service provider would be disconnected and would throw if you call into it.

I'm hoping we could solve this quite simply by dispatching to the PageContext's renderer's sync context, via the existing Dispatcher on that renderer. This should prevent anything like incoming navigation notifications arriving while the action has exclusive ownership of the sync context, but it would be good to confirm that by checking that our existing logic that invalidates PageContext also asserts that it's running on the sync context. Doing this also makes the name `Dispatch` make sense!

Also, do you think there's a case for supporting a `Func<IServiceProvider, Task>` overload, in case the developer wants to run async logic? And maybe a `Func<IServiceProvider, Task<T>>`? Our dispatch methods in other places do handle the `Action/Func<Task>/Func<Task<T>> ` combinations so it would only be a matter of passing the call through to the relevant overload on the renderer's dispatcher.

Note that if the developer does run async code within the dispatched callback, they would lose the guarantee that PageContext can't change. The guarantee would only hold until they yield the thread - when the continuation runs, the PageContext could have changed. I think that's OK, as the equivalent is already true for all Blazor component lifecycle methods.

Reply:
I took a look at the Renderer's dispatcher and that has only async APIs (which is probably fine, despite the note you gave).

Here's what it could look like:

```c#
    public async Task Dispatch(Action<IServiceProvider> action)
    {
        ArgumentNullException.ThrowIfNull(action);
        if (_currentPageContext is null)
        {
            throw new InvalidOperationException($"{nameof(Dispatch)} can't be called until after Blazor has started in the web view.");
        }

        await _currentPageContext.Renderer.Dispatcher.InvokeAsync(() => action(_currentPageContext.ServiceProvider));
    }
```

I think adding `Func<T>` overloads is fine. Easy to add, easy to support, and consistent with everywhere else.